File: programmers/C_language/대충만든자판.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

int* solution(
    const char* keymap[],
    size_t keymap_len,
    const char* targets[],
    size_t targets_len
) {

    // 알파벳 최소 클릭 수 저장
    int alpha[26];

    for(int i = 0; i < 26; i++) {
        alpha[i] = 101;
    }

    // keymap 최소값 저장
    for(int i = 0; i < keymap_len; i++) {

        for(int j = 0; keymap[i][j] != '\0'; j++) {

            int idx = keymap[i][j] - 'A';

            if(j + 1 < alpha[idx]) {
                alpha[idx] = j + 1;
            }
        }
    }

    // 결과 배열
    int* answer = (int*)malloc(sizeof(int) * targets_len);

    for(int i = 0; i < targets_len; i++) {

        int sum = 0;
        bool possible = true;

        for(int j = 0; targets[i][j] != '\0'; j++) {

            int idx = targets[i][j] - 'A';

            if(alpha[idx] == 101) {

                possible = false;

                break;
            }

            sum += alpha[idx];
        }

        answer[i] = possible ? sum : -1;
    }

    return answer;
}

File: Java/05_30/대충만든자판.java
class Solution {
    public int[] solution(String[] keymap, String[] targets) {

        int[] alpha = new int[26];

        // 초기값 설정
        for (int i = 0; i < 26; i++) {
            alpha[i] = 101;
        }

        // 각 알파벳의 최소 입력 횟수 저장
        for (int i = 0; i < keymap.length; i++) {
            for (int j = 0; j < keymap[i].length(); j++) {

                int idx = keymap[i].charAt(j) - 'A';

                if (j + 1 < alpha[idx]) {
                  //지금 발견한 입력횟수가 기존값보다 작으면 갱신
                    alpha[idx] = j + 1;
                }
            }
        }

        int[] answer = new int[targets.length];

        // targets 계산
        for (int i = 0; i < targets.length; i++) {

            int sum = 0;

            for (int j = 0; j < targets[i].length(); j++) {

                int idx = targets[i].charAt(j) - 'A';

                // 입력 불가능한 문자
                if (alpha[idx] == 101) {
                    sum = -1;
                    break;
                }

                sum += alpha[idx];
            }

            answer[i] = sum;
        }

        return answer;
    }
}